pkg/dialer/factory.go

```diff
@@ -56,6 +62,27 @@ func (f *Factory) ClusterDialer(clusterName string, retryOnError bool) (dialer.D
 	}, nil
 }
 
+func (f *Factory) ClusterDialHolder(clusterName string, retryOnError bool) (*transport.DialHolder, error) {
+	f.dialHoldersLock.RLock()
+	cached, ok := f.dialHolders[clusterName]
+	f.dialHoldersLock.RUnlock()
+	if ok {
+		return cached, nil
+	}
+
+	clusterDialer, err := f.ClusterDialer(clusterName, retryOnError)
```

Comment:
You are right, nice catch. I've updated it to prevent that specific case, thanks!

> And possibly garbage-collected after use

That wouldn't happen, since the problem is that client-go only ever adds to their transport caches, so it would never be freed.